processing/src/main/java/org/apache/druid/segment/loading/AcquireSegmentAction.java

```diff
@@ -0,0 +1,148 @@
+      long timeoutAt
+  )
+  {
+    final Closer safetyNet = Closer.create();
+    Throwable failure = null;
+
+    // getting the future kicks off any background action, so materialize them all to a list to get things started
+    final List<Future<Optional<Segment>>> futures = new ArrayList<>(acquireSegmentActions.size());
+    for (AcquireSegmentAction acquireSegmentAction : acquireSegmentActions) {
+      safetyNet.register(acquireSegmentAction);
+      // if we haven't failed yet, keep collecing futures (we always want to collect the actions themselves though
+      // to close
+      if (failure == null) {
+        try {
+          futures.add(acquireSegmentAction.getSegmentFuture());
+        }
+        catch (Throwable t) {
+          failure = t;
+        }
+      } else {
+        futures.add(Futures.immediateFuture(Optional.empty()));
+      }
+    }
+
+
+    final List<SegmentReference> segmentReferences = new ArrayList<>(acquireSegmentActions.size());
+    for (int i = 0; i < acquireSegmentActions.size(); i++) {
+      // if anything fails, want to ignore it initially so we can collect all additional futures to properly clean up
+      // all references before rethrowing the error
+      try {
+        final AcquireSegmentAction action = acquireSegmentActions.get(i);
+        final Future<Optional<Segment>> future = futures.get(i);
+        final Optional<Segment> segment = future.get(timeoutAt - System.currentTimeMillis(), TimeUnit.MILLISECONDS)
+                                                .map(safetyNet::register);
+        segmentReferences.add(
+            new SegmentReference(
+                action.getDescriptor(),
+                segmentMapFunction.apply(segment),
+                action
+            )
+        );
+      }
+      catch (Throwable t) {
+        if (failure == null) {
+          failure = t;
+        } else {
+          failure.addSuppressed(t);
+        }
+      }
+    }
+    if (failure != null) {
+      throw CloseableUtils.closeAndWrapInCatch(failure, safetyNet);
+    }
+    return segmentReferences;
+  }
+
+  public static AcquireSegmentAction missingSegment(final SegmentDescriptor descriptor)
+  {
+    return new AcquireSegmentAction(descriptor, () -> Futures.immediateFuture(Optional.empty()), NOOP_CLEANUP);
+  }
+
+  private final SegmentDescriptor segmentDescriptor;
+  private final Supplier<ListenableFuture<Optional<Segment>>> segmentFutureSupplier;
+  private final Closeable loadCleanup;
+
+  public AcquireSegmentAction(
+      SegmentDescriptor segmentDescriptor,
+      Supplier<ListenableFuture<Optional<Segment>>> segmentFutureSupplier,
+      Closeable loadCleanup
+  )
+  {
+    this.segmentDescriptor = segmentDescriptor;
+    this.segmentFutureSupplier = segmentFutureSupplier;
+    this.loadCleanup = loadCleanup;
+  }
+
+  public SegmentDescriptor getDescriptor()
+  {
+    return segmentDescriptor;
+  }
+
+  public ListenableFuture<Optional<Segment>> getSegmentFuture()
```

Comment:
Seems that in some cases, calling this method actually triggers the download to start. That isn't totally obvious so it should be called out in javadoc here.